java-checks/src/main/java/org/sonar/java/checks/OneExpectedCheckExceptionCheck.java

```diff
@@ -0,0 +1,167 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.java.model.ExpressionUtils;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LambdaExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+
+@Rule(key = "S5783")
+public class OneExpectedCheckExceptionCheck extends IssuableSubscriptionVisitor {
+
+  private static final MethodMatchers ASSERT_THROWS_MATCHER = MethodMatchers.create()
+    .ofTypes("org.junit.Assert", "org.junit.jupiter.api.Assertions")
+    .names("assertThrows")
+    .withAnyParameters()
+    .build();
+
+  private static final MethodMatchers JUNIT_FAIL_MATCHER = MethodMatchers.create()
+    .ofTypes("org.junit.Assert", "org.junit.jupiter.api.Assertions")
+    .names("fail")
+    .withAnyParameters()
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Arrays.asList(Tree.Kind.TRY_STATEMENT, Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) tree;
+      if (mit.arguments().size() >= 2 && ASSERT_THROWS_MATCHER.matches(mit)) {
+        Optional<IdentifierTree> expectedException = getExpectedException(mit.arguments().get(0));
+        ExpressionTree executable = mit.arguments().get(1);
```

Comment:
JUnit4 also supports assertions with messages - unfortunately as first parameter, different to JUnit5 (e.g. https://junit.org/junit4/javadoc/latest/src-html/org/junit/Assert.html#line.998). By relying on the exception type to be in position 0 and the executable in position 1, I'd argue it could come to FNs when a developer uses that JUnit4 assertion. Do we want to support that case as well?

Reply:
Good point!
JUnit 5 also has a message, but as the last parameter, everything works as expected.
And indeed, it is first in JUnit 4, I believe we should also support this case.